drizzlepac/haputils/catalog_utils.py

```diff
@@ -1414,111 +1419,165 @@ def identify_sources(self, **pars):
-                        log.info("")
-                        log.warning("Revert to using the original segmentation image and kernel.")
+                    # Background types: zero_background, sigma_clipped_background, twod_background
+                    # Compute a twod_background
+                    if (self.image.bkg_type.lower().startswith('sigma')):
 
-                        self.kernel = g2d_kernel
-                        segm_img = copy.deepcopy(custom_segm_img)
-                        del custom_segm_img
-                    # Use the RickerWavelet solution
+                        log.info("Recomputing the background image from a sigma-clipped background to a Background2D.")
+
+                        # In order to force the use of a background2D, some configuration values will be
+                        # re-set (i.e., bkg_skew_threshold and negative_percent).
+                        self.image.compute_background(self.param_dict['bkg_box_size'],
+                                                      self.param_dict['bkg_filter_size'],
+                                                      bkg_skew_threshold=0.0,
+                                                      negative_percent=100.0)
+
+                        if self.diagnostic_mode:
+                            outname = self.imgname.replace(".fits", "_bkg1.fits")
+                            fits.PrimaryHDU(data=self.image.bkg_background_ra).writeto(outname)
+
+                        # Need to remake image kernel as it has a dependence on self.bkg_rms_ra
+                        self.image.build_kernel(self.param_dict['bkg_box_size'],
+                                                self.param_dict['bkg_filter_size'],
+                                                self.param_dict['dao']['TWEAK_FWHMPSF'])
+
+
+                        sigma_for_threshold = self._nsigma
+                        rw2d_sigma_for_threshold = self._rw2d_nsigma
+
+                    # Re-compute a background2D with a higher threshold by increasing the nsigma used
+                    elif (self.image.bkg_type.lower().startswith('twod')):
+                        log.info("Increasing the threshold image (bkg + nsigma * 2.0) for improved source detection.")
+                        sigma_for_threshold = self._nsigma * 2.0
+                        rw2d_sigma_for_threshold = self._rw2d_nsigma * 2.0
+
+                    # Detect segments and evaluate the detection in terms of big islands or crowded fields
+                    # Round 2
+                    ncount += 1
+                    log.info("")
+                    log.info("With alternate background...using Gaussian kernel to generate a segmentation map.")
+                    del g_segm_img
+                    g_segm_img, g_is_big_crowded, g_bs, g_sf = self.detect_and_eval_segments(imgarr,
+                                                                                             g2d_kernel,
```

Comment:
The kernel was recomputed on line 1492 when calling 'build_kernel' again.  Did this redefine the kernel pointed to by 'g2d_kernel' or does 'g2d_kernel' need to be reset based on the new results from line 1492?  The code assumes it got redefined in place.

Reply:
Excellent catch!  Indeed the g2d_kernel _**and**_ the rw2d_kernel, need to be reset when a new background type is computed.